execution/model/operator.go

```diff
@@ -14,6 +15,44 @@ import (
 	"github.com/thanos-io/promql-engine/query"
 )
 
+type metadataKey string
+
+const key metadataKey = "promql-metadata"
+
+func AddMetadataStorage(ctx context.Context) context.Context {
+	return context.WithValue(ctx, key, &FanoutMetadata{
+		storage: make(map[string]map[string]any),
+	})
+}
+
+func GetMetadataStorage(ctx context.Context) *FanoutMetadata {
+	v := ctx.Value(key)
+	if v == nil {
+		return nil
+	}
+	return v.(*FanoutMetadata)
+}
+
+type FanoutMetadata struct {
```

Comment:
I think this could just be "Metadata" as its just a map; applications outside could build on top of this map since "fanout" is somewhat a thanos concept that this engine doesnt know about; its more an implementation detail of Select. Thanos could use this to just add fan out metadata in its Select implementation but no need that this library knows what this key contains really.